datafusion/physical-plan/src/joins/cross_join.rs

```diff
@@ -374,64 +376,147 @@ impl Stream for CrossJoinStream {
-                    let result =
-                        build_batch(self.left_index, &batch, left_data, &self.schema);
-                    self.join_metrics.input_batches.add(1);
-                    self.join_metrics.input_rows.add(batch.num_rows());
-                    if let Ok(ref batch) = result {
-                        join_timer.done();
-                        self.join_metrics.output_batches.add(1);
-                        self.join_metrics.output_rows.add(batch.num_rows());
-                    }
-                    self.left_index = 1;
-
-                    let mut right_batch = self.right_batch.lock();
-                    *right_batch = Some(batch);
-
-                    Some(result)
+            Some(Ok(right_batch)) => {
+                // Update the metrics.
+                self.join_metrics.input_batches.add(1);
+                self.join_metrics.input_rows.add(right_batch.num_rows());
+                if right_batch.num_rows() == 0 {
+                    return Poll::Ready(Ok(StatefulStreamResult::Continue));
                 }
-                other => other,
-            })
+                // New batch arrives, reset the indices.
+                self.left_batch_index = 0;
+                self.right_row_index = 0;
+                // Store the new batch into the state.
+                self.right_batch = right_batch;
+                self.state = CrossJoinStreamState::GenerateResult;
+                Poll::Ready(Ok(StatefulStreamResult::Continue))
+            }
+            Some(Err(err)) => Poll::Ready(Err(err)),
+        }
+    }
+
+    /// If there is non-paired rows in the probe batch, the function process them.
+    /// If not, it directs the state to fetching probe side.
+    fn generate_result(&mut self) -> Result<StatefulStreamResult<Option<RecordBatch>>> {
+        if self.right_row_index < self.right_batch.num_rows() {
+            // Right batch has some unpaired rows, continue with the next row.
+            let result_batch = self.build_batch()?;
+            Ok(StatefulStreamResult::Ready(Some(result_batch)))
+        } else {
+            self.state = CrossJoinStreamState::FetchProbeBatch;
+            Ok(StatefulStreamResult::Continue)
+        }
+    }
+
+    /// This function constructs a new `RecordBatch` by joining the left and right batches
+    /// based on the current indices. It also updates the metrics.
+    ///
+    /// # Arguments
+    /// * `self.left_data` - Array of `RecordBatch`es from the left side to be joined.
+    /// * `self.right_batch` - The current `RecordBatch` from the right side to be joined.
+    /// * `self.left_batch_index` - Index of the current left batch being processed.
+    /// * `self.right_row_index` - Index of the current row in the right batch to be joined.
+    /// * `join_metrics` - Metrics container to track performance of the join operation.
+
+    fn build_batch(&mut self) -> Result<RecordBatch> {
```

Comment:
minor: Got the idea, but it may be a bit confusing that docstring doesn't match signature.

BTW won't it be cleaner to leave it as a standalone function (as it was before) and leave mertics & offsets updates to `generate_result`?

Reply:
The function uses a few field from self object, so I think leaving it in the struct impl is better. But I move the metric and index updates to `generate_result` as you suggested.